### PR TITLE
Use angular-local-storage module  to better support local/session stor…

### PR DESCRIPTION
**Pull request by redwood-egarcea**
#457 Use angular-local-storage module  to better support local/session storage on safari while private browsing.

### CODE REVIEW

File: satellizer.js [hunk 1]
```diff
@@ -852,8 +819,11 @@
           }
         };
       }])
-    .config(['$httpProvider', function($httpProvider) {
-      $httpProvider.interceptors.push('satellizer.interceptor');
+      .config(['$httpProvider', 'satellizer.config', 'localStorageServiceProvider', function($httpProvider, config, localStorageServiceProvider) {
```
Comment (jderusse) [inline satellizer.js]: remove 2 spaces to respect indentation

File: satellizer.js [hunk 2]
```diff
@@ -786,42 +785,10 @@
...
-            return {
-              get: function(key) { return undefined; },
-              set: function(key, value) { return undefined; },
-              remove: function(key) { return undefined; }
-            };
-          }
-          break;
-      }
+    .factory('satellizer.storage', ['satellizer.config', 'localStorageService', function(config, localStorageService) {
+      angularLocalStorage.setStorageType(config.storage);
```
Comment (jderusse) [inline satellizer.js]: `storageType` is sets in the `config` step

File: satellizer.js [hunk 3]
```diff
@@ -852,8 +819,11 @@
           }
         };
       }])
-    .config(['$httpProvider', function($httpProvider) {
-      $httpProvider.interceptors.push('satellizer.interceptor');
+      .config(['$httpProvider', 'satellizer.config', 'localStorageServiceProvider', function($httpProvider, config, localStorageServiceProvider) {
+        $httpProvider.interceptors.push('satellizer.interceptor');
+        localStorageServiceProvider.prefix = config.tokenPrefix;
+        localStorageServiceProvider.storageType = config.storage;
+        angularLocalStorage = localStorageServiceProvider;
```
Comment (jderusse) [inline satellizer.js]: do not use global variables

File: satellizer.js [hunk 4]
```diff
@@ -852,8 +819,11 @@
           }
         };
       }])
-    .config(['$httpProvider', function($httpProvider) {
-      $httpProvider.interceptors.push('satellizer.interceptor');
+      .config(['$httpProvider', 'satellizer.config', 'localStorageServiceProvider', function($httpProvider, config, localStorageServiceProvider) {
+        $httpProvider.interceptors.push('satellizer.interceptor');
+        localStorageServiceProvider.prefix = config.tokenPrefix;
```
Comment (jderusse) [inline satellizer.js]: use setters as defined in the official documentation https://github.com/grevory/angular-local-storage#setprefix

```
.config(function (localStorageServiceProvider) {
  localStorageServiceProvider.setPrefix(config.tokenPrefix);
  localStorageServiceProvider.setStorageType(config.storage);
});
```

File: satellizer.js [hunk 5]
```diff
@@ -786,42 +785,10 @@
...
-            console.warn('Warning: Session Storage is disabled or unavailable. Satellizer will not work correctly.');
-            return {
-              get: function(key) { return undefined; },
-              set: function(key, value) { return undefined; },
-              remove: function(key) { return undefined; }
-            };
-          }
-          break;
-      }
+    .factory('satellizer.storage', ['satellizer.config', 'localStorageService', function(config, localStorageService) {
```
Comment (jderusse) [inline satellizer.js]: Given `satellizer.storage` and `localStorageService` use the same interface (`get`, `set`, `remove`), you can remove this service. 
If you want to keep the service `satellizer.storage` and restrict his scope, you can provide such service:

```
.factory('satellizer.storage', ['localStorageService', function(localStorageService) {
  return {
    get: localStorageService.get,
    set: localStorageService.set,
    remove: localStorageService.remove,
  };
}
```

**Comment by sahat** [comment]
Thanks @jderusse and @redwood-egarcea for PR and code review. I will merge it and release it along with `v0.12`.

**Comment by sahat** [comment]
Does it mean that setting the following line inside `config` block will set the storage type for the entire lifecycle of the application?

```js
localStorageServiceProvider.setStorageType(config.storage);
```

One of the requirements that someone else requested was so that you could switch between localStorage and sessionStorage on the fly. And AFAIK this local storage module does not support that.

**Comment by sahat** [comment]
Here is the code without pulling an entire library. It is even simpler than before (no more switch/case). 

I am still not really sure what to do when local storage is unavailable. Should warning message be displayed? Should it throw an error? Should it send a broadcast message on `$rootScope`? In any case, at the very least this should solve the DOM 22 Exception in Safari.

```js
.factory('satellizer.storage', ['$window', 'satellizer.config', function($window, config) {
      var browserSupportsLocalStorage = (function() {
        try {
          var supported = config.storageType in $window && $window[config.storageType] !== null;

          if (supported) {
            var key = Math.random().toString(36).substring(7);
            $window[config.storageType].setItem(key, '');
            $window[config.storageType].removeItem(key);
          }

          return supported;
        } catch (e) {
          return false;
        }
      })();

      if (!browserSupportsLocalStorage) {
        console.warn('Satellizer Warning: ' + config.storageType + ' is not available.');
      }

      return {
        get: function(key) {
          return browserSupportsLocalStorage ? $window[config.storageType].getItem(key) : undefined;
        },
        set: function(key, value) {
          return browserSupportsLocalStorage ? $window[config.storageType].setItem(key, value) : undefined;
        },
        remove: function(key) {
          return browserSupportsLocalStorage ? $window[config.storageType].removeItem(key): undefined;
        }
      };
    }])
```

**Comment by redwood-egarcea** [comment]
My opinion is that when local-storage is unavailable, the library should degrade to using cookies, similar to how angular-local-storage does. Safari is technically meeting the w3c web storage specification so to not support the most popular mobile browser is not an option in many cases.

**Comment by sahat** [comment]
Resolved in [0.12](https://github.com/sahat/satellizer/releases/tag/0.12.0) release.

**Pull request closed without merging**